### PR TITLE
[wip] app apply

### PR DESCRIPTION
**Pull request by lianxmfor**
- [x] add apply processing framework
- [x] implement apply entity
- [x] implement apply group
- [x] implement apply feature
- [x] parse a YAML file with multiple source
- [ ] add unit-test
- [ ] add integration-test

### CODE REVIEW

File: pkg/oomstore/apply.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+				EntityID:       entityID,
+				NewDescription: newEntity.Description,
+			}); err != nil {
+				return err
+			}
+		}
+	}
+
+	if len(newEntity.BatchFeatures) != 0 {
+		for _, batchFatures := range newEntity.BatchFeatures {
```
Comment (YikSanChan) [inline pkg/oomstore/apply.go:106]: batchFature -> batchFeature

File: pkg/oomstore/apply_test.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+     db-type-value: int
+stream-features:
+- name: c
+  db-type-value: xxx
+`
+	store := oomstore.TEST__New(nil, nil, nil)
+
+	require.Nil(t, store.Apply(context.Background(), apply.ApplyOpt{
+		R: strings.NewReader(entityYamlData),
+	}))
```
Comment (YikSanChan) [inline pkg/oomstore/apply_test.go:42]: Shall we add more tests?

File: pkg/oomstore/apply.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+	"gopkg.in/yaml.v3"
+)
+
+func (s *OomStore) Apply(ctx context.Context, opt apply.ApplyOpt) error {
+	data := make(map[string]interface{})
+	if err := yaml.NewDecoder(opt.R).Decode(data); err != nil {
+		return err
+	}
+
+	var kind string
```
Comment (YikSanChan) [inline pkg/oomstore/apply.go:21]: ## improvement

Shall we add a validation stage before starting to apply? In that stage, we make sure the input config file is valid.

## proposal

To make this cleaner, I wonder if we can parse the input file into a struct

```go
type ApplyGoal struct {
  ...
}
```

This struct contains ALL data we know about the input file. Then, we build a validation method `validate` against the `ApplyGoal` struct, call this method to ensure its validity.

Then, we build a `apply(ApplyGoal)` method that does the apply.

## benefits

More declarative and less procedural, more readable

## cons

Less straightforward

@lianxmfor @jinghancc @wfxr what do you think?

File: pkg/oomstore/apply.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+		return s.metadata.WithTransaction(ctx, func(c context.Context, tx metadata.WriteStore) error {
+			return s.applyGroup(ctx, tx, group)
+		})
+	default:
+		return fmt.Errorf("invalid kind %s", kind)
+	}
+}
+
+func (s *OomStore) applyEntity(ctx context.Context, txStore metadata.WriteStore, newEntity apply.Entity) error {
```
Comment (YikSanChan) [inline pkg/oomstore/apply.go:71]: txStore -> store? Since the method allows non tx store to be passed in

File: pkg/oomstore/apply_test.go (hunk 2 above)
Comment (lianxmfor) [reply pkg/oomstore/apply_test.go:42]: Yes, I am about to add more tests

File: pkg/oomstore/apply.go (hunk 4 above)
Comment (lianxmfor) [reply pkg/oomstore/apply.go:71]: Not sure I understand...  but here the methods that require transactions are called through the txStore  and the non-tx methods are not called using the txStore

Comment (YikSanChan) [reply pkg/oomstore/apply.go:71]: 🆗

File: pkg/oomstore/apply.go (hunk 3 above)
Comment (lianxmfor) [reply pkg/oomstore/apply.go:21]: sound good! But what do I need to validate? I'm currently thinking of the kind field

Comment (YikSanChan) [reply pkg/oomstore/apply.go:21]: Hmm maybe as long as the parsing works, then there is little rest for validation.

File: pkg/oomstore/apply.go (hunk 1 above)
Comment (lianxmfor) [reply pkg/oomstore/apply.go]: fixed

File: pkg/oomstore/apply.go (hunk 3 above)
Comment (lianxmfor) [reply pkg/oomstore/apply.go]: I have refactored according to this improvement plan, but I think the code is a bit ugly...  @YikSanChan @jinghancc @wfxr

File: featctl/cmd/apply.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+kind: Entity
+name: user
+length: 8
+description: a description
+batch-features:
+- group: device
+  description: a description
+  features:
+  - name: model
+    db-type-value: varchar(16)
```
Comment (jinghancc) [inline featctl/cmd/apply.md:17]: typo: db_value_type?

File: pkg/oomstore/apply.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	}
+
+	return s.metadata.WithTransaction(ctx, func(c context.Context, tx metadata.WriteStore) error {
+		var (
+			entityMp = make(map[string]int) // entity name -> entity id
+			groupMp  = make(map[string]int) // group name -> group id
+		)
+
+		// apply entity
+		for _, entity := range stage.NewEntitys {
```
Comment (YikSanChan) [inline pkg/oomstore/apply.go:30]: fix all entitys -> entities

File: pkg/oomstore/types/apply/apply.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+type ApplyOpt struct {
+	R io.Reader
+}
+
+type Feature struct {
+	Kind        string `mapstructure:"kind"`
+	Name        string `mapstructure:"name"`
+	GroupName   string `mapstructure:"group-name"`
+	DBValueType string `mapstructure:"db-type-value"`
```
Comment (jinghancc) [inline pkg/oomstore/types/apply/apply.go:28]: `db-value-type` or `db_value_type`?

Comment (lianxmfor) [reply pkg/oomstore/types/apply/apply.go:28]: hmm, I am not sure which is better... The - is currently used uniformly

File: pkg/oomstore/apply.go (hunk 6 above)
Comment (lianxmfor) [reply pkg/oomstore/apply.go]: done

File: pkg/oomstore/apply.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+		return nil
+	})
+}
+
+func (s *OomStore) applyEntity(ctx context.Context, txStore metadata.WriteStore, newEntity apply.Entity) (int, error) {
+	entityExist := true
+
+	entity, err := s.metadata.GetEntityByName(ctx, newEntity.Name)
+	if err != nil {
+		if err.Error() != fmt.Sprintf("feature entity '%s' not found", newEntity.Name) {
```
Comment (YikSanChan) [inline pkg/oomstore/apply.go:77]: make "not found" an error type

**Review by YikSanChan — APPROVED** [review]
LGTM

File: pkg/oomstore/types/options.go [hunk 9]
```diff
@@ -81,3 +81,7 @@ type UpdateGroupOpt struct {
 type SyncOpt struct {
 	RevisionID int
 }
+
+type ApplyOpt struct {
+	Filepath string
```
Comment (jinghancc) [inline pkg/oomstore/types/options.go:86]: typo: FilePath

Comment (lianxmfor) [reply pkg/oomstore/types/options.go]: fixed

**Review by jinghancc — APPROVED** [review]
LGTM

**Comment by lianxmfor** [comment]
I will do error type check and unit-test, integration-test in other PR

**Pull request merged**